Remotion/Data/DomainObjects.UnitTests/Linq/IntegrationTests/TableValuedParameterIntegrationTest.cs

```diff
@@ -0,0 +1,318 @@
+        from o in QueryFactory.CreateLinqQuery<Order>()
+        where possibleItems.Contains((Guid)o.ID.Value)
+        select o;
+
+    CheckQueryResult(orders, []);
+  }
+
+  [Test]
+  public void Contains_WithCollectionOfT ()
+  {
+    var possibleItems = SetupCollection<ICollection<Guid>>([ (Guid)DomainObjectIDs.Order1.Value, (Guid)DomainObjectIDs.Order3.Value, (Guid)DomainObjectIDs.Order3.Value ]);
+
+    var orders =
+        from o in QueryFactory.CreateLinqQuery<Order>()
+        where possibleItems.Contains((Guid)o.ID.Value)
+        select o;
+
+    CheckQueryResult(orders, DomainObjectIDs.Order1, DomainObjectIDs.Order3);
+  }
+
+  [Test]
+  public void Contains_WithReadOnlyCollectionOfT ()
+  {
+    var possibleItems = SetupCollection<IReadOnlyCollection<Guid>>([ (Guid)DomainObjectIDs.Order1.Value, (Guid)DomainObjectIDs.Order3.Value, (Guid)DomainObjectIDs.Order3.Value ]);
+
+    var orders =
+        from o in QueryFactory.CreateLinqQuery<Order>()
+        where possibleItems.Contains((Guid)o.ID.Value)
+        select o;
+
+    CheckQueryResult(orders, DomainObjectIDs.Order1, DomainObjectIDs.Order3);
+  }
+
+  [Test]
+  public void Contains_WithPlainCollection ()
+  {
+    var possibleItems = (IEnumerable<Guid>)SetupCollection<ICollection>([ (Guid)DomainObjectIDs.Order1.Value, (Guid)DomainObjectIDs.Order3.Value, (Guid)DomainObjectIDs.Order3.Value ]);
+
+    var orders =
+        from o in QueryFactory.CreateLinqQuery<Order>()
+        where possibleItems.Contains((Guid)o.ID.Value)
+        select o;
+
+    CheckQueryResult(orders, DomainObjectIDs.Order1, DomainObjectIDs.Order3);
+  }
+
+  [Test]
+  public void Contains_WithSetOfT ()
+  {
+    var possibleItems = SetupCollection<ISet<Guid>>([(Guid)DomainObjectIDs.Order1.Value, (Guid)DomainObjectIDs.Order3.Value]);
+
+    var orders =
+        from o in QueryFactory.CreateLinqQuery<Order>()
+        where possibleItems.Contains((Guid)o.ID.Value)
+        select o;
+
+    CheckQueryResult(orders, DomainObjectIDs.Order1, DomainObjectIDs.Order3);
+  }
+
+  [Test]
+  public void Contains_WithSetOfT_UsesTableTypeWithUniqueConstraint ()
+  {
+    // set up a set with duplicates to get a database exception when we insert these duplicates into a table with unique constraint
+    var possibleItems = SetupCollection<ISet<Guid>>([(Guid)DomainObjectIDs.Order1.Value, (Guid)DomainObjectIDs.Order3.Value, (Guid)DomainObjectIDs.Order3.Value]);
+
+    var orders =
+        from o in QueryFactory.CreateLinqQuery<Order>()
+        where possibleItems.Contains((Guid)o.ID.Value)
+        select o;
+
+    Assert.That(() => CheckQueryResult(orders, DomainObjectIDs.Order1, DomainObjectIDs.Order3),
+        Throws.InstanceOf<RdbmsProviderException>()
+            .With.Message.Contains("Error while executing SQL command: Violation of UNIQUE KEY constraint")
+            .With.Message.Contains("Cannot insert duplicate key in object 'dbo.@1'. The duplicate key value is (83445473-844a-4d3f-a8c3-c27f8d98e8ba)."));
```

Comment:
just a note: this might one day break due to changes in sql server messages. if it does, we might want to look into checking less details in the message.